pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go

```diff
@@ -131,191 +168,534 @@ func CreateIndex(b BuildCtx, n *tree.CreateIndex) {
 			panic(pgerror.Newf(pgcode.DuplicateRelation, "index with name %q already exists", n.Name))
 		}
 	}
+	// Assign the ID here, since we may have added columns
+	// and made a new primary key above.
+	idxSpec.secondary.SourceIndexID = idxSpec.primary.IndexID
+	idxSpec.secondary.IndexID = nextRelationIndexID(b, relation)
+	idxSpec.secondary.TemporaryIndexID = idxSpec.secondary.IndexID + 1
+	// Add columns for the secondary index.
+	addColumnsForSecondaryIndex(b, n, relation, &idxSpec)
+	// If necessary set up the partitioning descriptor for the index.
+	maybeAddPartitionDescriptorForIndex(b, n, &idxSpec)
+	// If necessary setup a partial predicate
+	maybeAddIndexPredicate(b, n, &idxSpec)
+	// Picks up any geoconfig parameters, hash sharded one are
+	// picked independently.
+	maybeApplyStorageParameters(b, n, &idxSpec)
+	keyIdx := 0
+	keySuffixIdx := 0
+	for _, ic := range idxSpec.columns {
+		if ic.Kind == scpb.IndexColumn_KEY {
+			ic.OrdinalInKind = uint32(keyIdx)
+			keyIdx++
+		} else if ic.Kind == scpb.IndexColumn_KEY_SUFFIX {
+			ic.OrdinalInKind = uint32(keySuffixIdx)
+			keySuffixIdx++
+		}
+	}
+	idxSpec.data = &scpb.IndexData{TableID: idxSpec.secondary.TableID, IndexID: idxSpec.secondary.IndexID}
+	idxSpec.primary = nil
+	idxSpec.apply(b.Add)
+	// Apply the name once everything else has been created, since we need
+	// elements to resolve things.
+	indexName := string(n.Name)
+	if indexName == "" {
+		_, _, tbl := scpb.FindTable(relationElements)
+		numImplicitColumns := 0
+		if idxSpec.partitioning != nil {
+			numImplicitColumns = int(idxSpec.partitioning.NumImplicitColumns)
+		}
+		indexName = getImplicitSecondaryIndexName(b,
+			tbl,
+			idxSpec.secondary.IndexID,
+			numImplicitColumns,
+		)
+	}
+	b.Add(&scpb.IndexName{
+		TableID: idxSpec.secondary.TableID,
+		IndexID: idxSpec.secondary.IndexID,
+		Name:    indexName,
+	})
+	// Construct the temporary objects from the index spec, since these will
+	// be transient.
+	tempIdxSpec.temporary = &scpb.TemporaryIndex{
+		Index:                    protoutil.Clone(idxSpec.secondary).(*scpb.SecondaryIndex).Index,
+		IsUsingSecondaryEncoding: true,
+	}
```

Comment:
Do yourself a favour and use `idxSpec.clone()` or, even better, introduce a `makeTempIndexSpec(src indexSpec) (tmp indexSpec)` constructor and use that both there and in `makeSwapIndexSpec`.